Adds services for pruning roles when requesting remote resources.

tcsc opened the pull request:
Currently, when a user creates an Acess Request for a resource in a remote cluster
Teleport lacks the appropriate information to decide which roles to offer and falls
back to suggesting all available `search_as` roles for the reequesting user.

This patch provides the back-end services to allow Teleport to prune the roles for
remote-resource requests in the same way as requests for local resources.

For context, the entire remote pruning process will look something like this:

```mermaid
sequenceDiagram
    actor rui as root UI
    participant rp as root cluster proxy
    participant ra as root cluster auth
    participant la as leaf cluster auth

    rui ->> rp: GET /accessrequestroles?...
    alt request has local (root) cluster resources
        rp ->> ra: GetAccessCapabilities()
        ra ->> rp: request_roles
    else request has leaf cluster resources
        note over rp: Root cluster collects all user<br/>`SearchAs` roles
        rp ->> la: GetRemoteAccessCapabilities(remote_search_as_roles: user_sa_roles)
        note over la: Map the supplied root-cluster search-as roles onto<br/>leaf-custer roles using the configured <br/>Trusted Cluster Role Map.
        note over la: Prune the mapped leaf-cluster roles using the same<br/>algorithm as for local resources, using the mapped<br/>roles for the Search-As input rather than collect<br/>them from the user's roleset.
        note over la: Re-map the pruned leaf-cluster roles back into a<br/>subset of the supplied root-cluster search-as roles<br/>using the Trusted Cluster Role Map. 
        la ->> rp: applicable_roles
    end
    rp ->> rui: applicable_roles
```

Also note that this change only includes includes the back-end changes that will be
executed on a leaf cluster. The front-end, root-cluster changes are coming in a
subsequent PR.


Addresses: #56315
Changelog: Added services to correctly choose Access Request roles in remote clusters.

tcsc commented:
Remote role pruning algorithm sequence
```mermaid
sequenceDiagram
    actor rui as root UI
    participant rp as root cluster proxy
    participant ra as root cluster auth
    participant la as leaf cluster auth

    rui ->> rp: GET /accessrequestroles?...
    alt request has local (root) cluster resources
        rp ->> ra: GetAccessCapabilities()
        ra ->> rp: request_roles
    else request has leaf cluster resources
        note over rp: Root cluster collects all user<br/>`SearchAs` roles
        rp ->> la: GetRemoteAccessCapabilities(remote_search_as_roles: user_sa_roles)
        note over la: Map the supplied root-cluster search-as roles onto<br/>leaf-custer roles using the configured <br/>Trusted Cluster Role Map.
        note over la: Prune the mapped leaf-cluster roles using the same<br/>algorithm as for local resources.
        note over la: Re-map the pruned leaf-cluster roles back into a<br/>subset of the supplied root-cluster search-as roles<br/>using the Trusted Cluster Role Map. 
        la ->> rp: applicable_roles
    end
    rp ->> rui: applicable_roles
```

nklaassen commented on lib/services/trustedcluster.go:
```diff
@@ -142,7 +144,50 @@ func MapRoles(r types.RoleMap, remoteRoles []string) ([]string, error) {
+	return index, nil
+}
+
+// MapRoles maps local roles to remote roles
```
doesn't it map remote roles to local roles?

nklaassen commented on lib/services/access_request_test.go:
```diff
@@ -1695,6 +1700,251 @@ func TestPruneRequestRoles(t *testing.T) {
 
+// requireBadParameter is a [require.ErrorAssertionFunc] that asserts the supplied error value is a BadParameter error
+func requireBadParameter(t require.TestingT, err error, msgAndArgs ...any) {
+	require.Error(t, err, msgAndArgs...)
+	require.True(t, trace.IsBadParameter(err), "Expected BadParameter, got %T: %s", err, err.Error())
```
perhaps
```suggestion
	require.ErrorAs(t, err, new(*trace.BadParameterError), msgAndArgs...)
```

nklaassen commented on lib/auth/auth.go:
```diff
@@ -6051,6 +6051,12 @@ func (a *Server) GetAccessCapabilities(ctx context.Context, req types.AccessCapa
 	return caps, nil
 }
 
+// GetRemoteAccessCapabilities requires information more readily available in
+// [ServerWithRoles], so the main implementation is there.
+func (a *Server) GetRemoteAccessCapabilities(context.Context, types.RemoteAccessCapabilitiesRequest) (*types.RemoteAccessCapabilities, error) {
+	return nil, trace.NotImplemented("GetRemoteAccessCapabilities not implemented")
+}
```
is this easily avoidable or does it really need to be added to the AuthService interface

fspmarshall reviewed (approved):
LGTM overall.  My one main suggestion would be to more thoroughly comment/document what remote pruning is doing and why (both in the protos and the business logic).  I think this is the kind of feature that can be particularly opaque when you first come across it.

tcsc committed:
Adds services for pruning roles when requesting remote resources.

Currently, when a user creates an Acess Request for a resource in a remote cluster
Teleport lacks the appropriate information to decide which roles to offer and falls
back to suggesting all available `search_as` roles for the reequesting user.

This patch provides the back-end services to allow Teleport to prune the roles for
remote-resource requests in the same way as requests for local resources.

Also note that this change only includes includes the back-end changes that will be
executed on a leaf cluster. The front-end, root-cluster changes are coming in a
subsequent PR.

Addresses: #56315
Changelog: Added services to correctly choose Access Request roles in remote clusters.
